### ISSUE TITLE
nuget: ThreadAbortException in SchedulerThread

### ISSUE THREAD

**Issue by rolfbjarne** [mono/mono#15418]
## Steps to Reproduce

Unfortunately I don't have a way to reproduce, I've seen it once when running "nuget restore" on a bot. It's not related to the particular solution/project that was restored.

### Current Behavior

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Failed to compare two elements in the array. ---> System.Threading.ThreadAbortException: Thread was being aborted.
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at System.Threading.Timer+TimerComparer.Compare (System.Threading.Timer tx, System.Threading.Timer ty) <0x1087381b0 + 0x00035> in <b7d51802a4dc4818a2e8326fef769990>:0 
  at (wrapper unbox) System.Threading.Timer+TimerComparer.Compare(System.Threading.Timer,System.Threading.Timer)
  at System.Collections.Generic.ArraySortHelper`1[T].PickPivotAndPartition (T[] keys, System.Int32 lo, System.Int32 hi, System.Comparison`1[T] comparer) [0x0003d] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].IntroSort (T[] keys, System.Int32 lo, System.Int32 hi, System.Int32 depthLimit, System.Comparison`1[T] comparer) [0x00069] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].IntroSort (T[] keys, System.Int32 lo, System.Int32 hi, System.Int32 depthLimit, System.Comparison`1[T] comparer) [0x00074] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].IntrospectiveSort (T[] keys, System.Int32 left, System.Int32 length, System.Comparison`1[T] comparer) [0x00013] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].Sort (T[] keys, System.Int32 index, System.Int32 length, System.Collections.Generic.IComparer`1[T] comparer) [0x0001c] in <b7d51802a4dc4818a2e8326fef769990>:0 
   --- End of inner exception stack trace ---
  at System.Collections.Generic.ArraySortHelper`1[T].Sort (T[] keys, System.Int32 index, System.Int32 length, System.Collections.Generic.IComparer`1[T] comparer) [0x00039] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Array.Sort[T] (T[] array, System.Int32 index, System.Int32 length, System.Collections.Generic.IComparer`1[T] comparer) [0x0004d] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.List`1[T].Sort (System.Int32 index, System.Int32 count, System.Collections.Generic.IComparer`1[T] comparer) [0x0002b] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Collections.Generic.List`1[T].Sort (System.Collections.Generic.IComparer`1[T] comparer) [0x00008] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.Timer+Scheduler.RunSchedulerLoop () [0x0001a] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.Timer+Scheduler.SchedulerThread () [0x00029] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <b7d51802a4dc4818a2e8326fef769990>:0 
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <b7d51802a4dc4818a2e8326fef769990>:0 
```

For the record: it's the failing test [here](https://dev.azure.com/xamarin/internal/_build/results?buildId=3312&view=ms.vss-test-web.build-test-results-tab).

### Expected Behavior

No unhandled exception.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

```
$ mono --version
Mono JIT compiler version 6.0.0.176 (2019-02/d75c142a464 Fri Apr 12 17:31:22 EDT 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(600)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

**Comment by lambdageek** [comment]
Seems like we're doing the moral equivalent of:

```csharp
using System;
using System.Threading;

public class Foo {
	public static void Main () {
		var t = new Thread (RunMe);
		t.Start ();
		Thread.Sleep (100);
		t.Abort ();
		t.Join ();
		Console.WriteLine ("bye bye");
		
	}

	public static void RunMe () {
		try {
			while (true);
		} catch (ThreadAbortException e) {
			Console.WriteLine ("caught tae");
			throw new InvalidOperationException ("ioe tae", e);
		} finally {
			Console.WriteLine ("Bye");
		}
	}
}
```

Which throws the InvalidOperationException on the thread being aborted (Linux, Mono 5.20.1.19).
<details>
<pre>
caught tae
Bye

Unhandled Exception:
System.InvalidOperationException: ioe tae ---> System.Threading.ThreadAbortException: Thread was being aborted.
  at Foo.RunMe () [0x00002] in <1d6606fc4788405da2461683fc070347>:0 
   --- End of inner exception stack trace ---
  at Foo.RunMe () [0x00021] in <1d6606fc4788405da2461683fc070347>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x7f8a9632a000 + 0x00096> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7f8a96328390 + 0x0012f> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7f8a96328340 + 0x00035> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x7f8a963282d0 + 0x0003e> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ThreadHelper.ThreadStart () <0x7f8a9632a140 + 0x00032> in <6649516e5b3542319fb262b421af0adb>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: ioe tae ---> System.Threading.ThreadAbortException: Thread was being aborted.
  at Foo.RunMe () [0x00002] in <1d6606fc4788405da2461683fc070347>:0 
   --- End of inner exception stack trace ---
  at Foo.RunMe () [0x00021] in <1d6606fc4788405da2461683fc070347>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x7f8a9632a000 + 0x00096> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7f8a96328390 + 0x0012f> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7f8a96328340 + 0x00035> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x7f8a963282d0 + 0x0003e> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.ThreadHelper.ThreadStart () <0x7f8a9632a140 + 0x00032> in <6649516e5b3542319fb262b421af0adb>:0 
</pre>
</details>

Things I don't know at this point:
1. What does .net framework do here? (will try when I have a Windows VM available)
2. Why the Timer-Scheduler thread is being aborted in nuget

**Comment by rolfbjarne** [comment]
> 2\. Why the Timer-Scheduler thread is being aborted in nuget

My guess would be because the process is exiting.

This happened at the very end of nuget's execution ([log](https://gist.github.com/rolfbjarne/f8d57f54762ec311074d51be59ba572e#file-nuget-log-L208)).

**Comment by lambdageek** [comment]
Instead of patching up the Timer-Scheduler code, we will add some `catch (ThreadAbortException){}` clauses all over [ArraySortHelper.cs](https://github.com/mono/corert/blob/master/src/System.Private.CoreLib/shared/System/Collections/Generic/ArraySortHelper.cs)

**Comment by lambdageek** [comment]
There's a beginnings of a test case (at least for `Array.Sort<T>(..., comparer)`) in a comment https://github.com/mono/mono/pull/15454#issuecomment-506535048

(And also note the following comment that an improved test would be just to make an `IComparer` that calls `Thread.Sleep(Timeout.Infinite);`)

**Comment by slluis** [comment]
@lambdageek @marek-safar This making the VS Mac unit test suite fail very often. Any chance this can be fixed? It would be very helpful.

**Comment by marek-safar** [comment]
@lambdageek could you add this to your threading issues list for this week?

**Comment by lambdageek** [comment]
Should be fixed in 6.12.0.102 (https://github.com/mono/mono/commit/af315f44c40dfb8767d64920bae2cdb8da7cc3f8) once the package builds.
